core/src/main/scala/akka/persistence/r2dbc/journal/JournalDao.scala

```diff
@@ -180,31 +183,54 @@ private[r2dbc] class JournalDao(journalSettings: R2dbcSettings, connectionFactor
       stmt
     }
 
-    val result = {
-
-      val insertSql =
-        if (useTimestampFromDb) insertEventWithTransactionTimestampSql
-        else insertEventWithParameterTimestampSql
-
+    if (useTimestampFromDb) {
       val totalEvents = events.size
-      if (totalEvents == 1)
-        r2dbcExecutor.updateOne(s"insert [$persistenceId]") { connection =>
-          bind(connection.createStatement(insertSql), events.head)
-        }
-      else
-        r2dbcExecutor.updateInBatch(s"batch insert [$persistenceId], [$totalEvents] events") { connection =>
-          events.foldLeft(connection.createStatement(insertSql)) { (stmt, write) =>
-            bind(stmt, write).add()
+      if (totalEvents == 1) {
+        val result = r2dbcExecutor.updateOneReturning(s"insert [$persistenceId]")(
+          connection => bind(connection.createStatement(insertEventWithTransactionTimestampSql), events.head),
+          row => Some(row.get(0, classOf[Instant])))
+        if (log.isDebugEnabled())
+          result.foreach { _ =>
+            log.debug("Wrote [{}] events for persistenceId [{}]", 1, events.head.persistenceId)
           }
+        result
+      } else {
+        val result = r2dbcExecutor.updateInBatch(s"batch insert [$persistenceId], [$totalEvents] events") {
+          connection =>
+            events.foldLeft(connection.createStatement(insertEventWithTransactionTimestampSql)) { (stmt, write) =>
+              bind(stmt, write).add()
+            }
         }
-    }
 
-    if (log.isDebugEnabled())
-      result.foreach { updatedRows =>
-        log.debug("Wrote [{}] events for persistenceId [{}]", updatedRows, events.head.persistenceId)
+        if (log.isDebugEnabled())
+          result.foreach { updatedRows =>
+            log.debug("Wrote [{}] events for persistenceId [{}]", updatedRows, events.head.persistenceId)
+          }
+
+        result.map(_ => None)(ExecutionContexts.parasitic)
```

Comment:
Returning should also work with batch inserts. Have you tried it? 

The annoying part is that it will return a list  of timestamps as large as the number of items in the batch. 

For example, 

```sql
postgres=# INSERT INTO foo(name) VALUES ('abc'), ('def') returning transaction_timestamp();

     transaction_timestamp
-------------------------------
 2022-01-12 13:10:41.451654+00
 2022-01-12 13:10:41.451654+00
(2 rows)

INSERT 0 2
```

Reply:
yup, changed to returning for the batch case also